src/daft-table/src/ops/joins/mod.rs

```diff
@@ -156,27 +162,49 @@ impl Table {
         right: &Self,
         left_on: &[ExprRef],
         right_on: &[ExprRef],
-        inner_join: impl Fn(&Table, &Table) -> DaftResult<(Series, Series)>,
+        join_fn: impl Fn(&Table, &Table) -> DaftResult<(Series, Series)>,
     ) -> DaftResult<Self> {
         let join_schema = infer_join_schema(&self.schema, &right.schema, left_on, right_on)?;
-        if self.is_empty() || right.is_empty() {
-            return Self::empty(Some(join_schema.into()));
-        }
         let ltable = self.eval_expression_list(left_on)?;
         let rtable = right.eval_expression_list(right_on)?;
 
         let (ltable, rtable) = match_types_for_tables(&ltable, &rtable)?;
-        let (lidx, ridx) = inner_join(&ltable, &rtable)?;
+        let (lidx, ridx) = join_fn(&ltable, &rtable)?;
+
         let mut join_fields = ltable
             .column_names()
             .iter()
             .map(|s| self.schema.get_field(s).cloned())
             .collect::<DaftResult<Vec<_>>>()?;
 
-        let mut join_series = self
-            .get_columns(ltable.column_names().as_slice())?
-            .take(&lidx)?
-            .columns;
+        let mut join_series = vec![];
+
+        for (l, r) in ltable
+            .column_names()
+            .iter()
+            .zip(rtable.column_names().iter())
+        {
+            if l == r {
+                let lcol = self.get_column(l)?;
+                let rcol = right.get_column(r)?;
+
+                let mut growable =
+                    make_growable(l, lcol.data_type(), vec![lcol, rcol], false, lcol.len());
+
+                for (li, ri) in lidx.u64()?.into_iter().zip(ridx.u64()?) {
```

Comment:
spoke offline about factoring this out just for Outer Join and using if_else(l.take, r.take)